src/EditorFeatures/Core/Implementation/KeywordHighlighting/AbstractKeywordHighlighter.cs

```diff
@@ -12,67 +13,86 @@ internal abstract class AbstractKeywordHighlighter<TNode> : AbstractKeywordHighl
     {
         protected sealed override bool IsHighlightableNode(SyntaxNode node) => node is TNode;
 
-        protected sealed override IEnumerable<TextSpan> GetHighlightsForNode(SyntaxNode node, CancellationToken cancellationToken)
-            => GetHighlights((TNode)node, cancellationToken);
+        protected sealed override void AddHighlightsForNode(SyntaxNode node, List<TextSpan> highlights, CancellationToken cancellationToken)
+            => AddHighlights((TNode)node, highlights, cancellationToken);
 
-        protected abstract IEnumerable<TextSpan> GetHighlights(TNode node, CancellationToken cancellationToken);
+        protected abstract void AddHighlights(TNode node, List<TextSpan> highlights, CancellationToken cancellationToken);
     }
 
     internal abstract class AbstractKeywordHighlighter : IHighlighter
     {
+        private static readonly ObjectPool<List<TextSpan>> s_textSpanListPool = new ObjectPool<List<TextSpan>>(() => new List<TextSpan>());
+        private static readonly ObjectPool<List<SyntaxToken>> s_tokenListPool = new ObjectPool<List<SyntaxToken>>(() => new List<SyntaxToken>());
+
         protected abstract bool IsHighlightableNode(SyntaxNode node);
 
-        public IEnumerable<TextSpan> GetHighlights(
-            SyntaxNode root, int position, CancellationToken cancellationToken)
+        public void AddHighlights(
+            SyntaxNode root, int position, List<TextSpan> highlights, CancellationToken cancellationToken)
         {
-            foreach (var token in GetTokens(root, position))
+            var _1 = s_textSpanListPool.GetPooledObject();
```

Comment:
please give these variables names instead of `_1` and `_2`

Reply:
done.  but i don't think it adds anything.  these variables aren't important.  they're just the temps that are created in order to dispose.  you should only be using the .Object inside of them.

Reply:
Simplified through a new extension method.